timemachine/cpp/src/rmsd_restraint.cu

```diff
@@ -173,52 +196,41 @@ void RMSDRestraint<RealType>::execute_device(
     // (ytz): taken from here https://j-towns.github.io/papers/svd-derivative.pdf
     // using equations 30-31. note that adjoints of the singular values are always zero,
     // so the middle term is skipped.
-    Eigen::MatrixXd v_a = v_a_t.transpose();
-    Eigen::MatrixXd lhs_1 = u*F.cwiseProduct(u.transpose()*u_a - u_a.transpose()*u)*smat;
-    Eigen::MatrixXd lhs_2 = (eye - u*u.transpose())*u_a*smat_inv;
-    Eigen::MatrixXd lhs = (lhs_1+lhs_2)*v.transpose();
-    Eigen::MatrixXd rhs_1 = smat*(F.cwiseProduct(v.transpose()*v_a - v_a.transpose()*v))*v.transpose();
-    Eigen::MatrixXd rhs_2 = smat_inv*v_a.transpose()*(eye - v*v.transpose());
-    Eigen::MatrixXd rhs = u*(rhs_1 + rhs_2);
-    Eigen::MatrixXd c_adjoint = lhs + rhs;
-    Eigen::MatrixXd x2_adjoint = (c_adjoint*x1.transpose()).transpose();
-    Eigen::MatrixXd x1_adjoint = x2*c_adjoint;
-
-    std::vector<unsigned long long> du_dx(N*3, 0);
-
-    for(int b=0; b < B; b++) {
-        int src_idx = h_atom_map_[b*2+0];
-        du_dx[src_idx*3+0] += static_cast<unsigned long long>(llrint(k_*x1_adjoint(b, 0)*FIXED_EXPONENT));
-        du_dx[src_idx*3+1] += static_cast<unsigned long long>(llrint(k_*x1_adjoint(b, 1)*FIXED_EXPONENT));
-        du_dx[src_idx*3+2] += static_cast<unsigned long long>(llrint(k_*x1_adjoint(b, 2)*FIXED_EXPONENT));
-
-        int dst_idx = h_atom_map_[b*2+1];
-        du_dx[dst_idx*3+0] += static_cast<unsigned long long>(llrint(k_*x2_adjoint(b, 0)*FIXED_EXPONENT));
-        du_dx[dst_idx*3+1] += static_cast<unsigned long long>(llrint(k_*x2_adjoint(b, 1)*FIXED_EXPONENT));
-        du_dx[dst_idx*3+2] += static_cast<unsigned long long>(llrint(k_*x2_adjoint(b, 2)*FIXED_EXPONENT));
-    }
-
-    gpuErrchk(cudaMemcpy(d_u_buf_, &nrg, sizeof(*d_u_buf_), cudaMemcpyHostToDevice));
-    gpuErrchk(cudaMemcpy(d_du_dx_buf_, &du_dx[0], N*3*sizeof(*d_du_dx_buf_), cudaMemcpyHostToDevice));
-
-    const int blocks= (N_+32-1)/32;
-    const int tpb = 32;
-
-    k_finalize<<<blocks, tpb, 0, stream>>>(
-        N,
-        d_du_dx_buf_,
-        d_u_buf_,
+    Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic> v_a = v_a_t.transpose();
```

Comment:
I think some of these expression can be optimized a bit here more (eg. v_a_t.transpose() = v_a), but we can do this in a later pass if you'd like.

Reply:
Happy to do that in this PR, especially since I will probably wait until you merge your updates to this logic anyways.

Reply:
Done in latest